### DISCUSSION TITLE
Keep queue running during lifetime of application

### DISCUSSION THREAD

**Discussion by vsoch** [riverqueue/river#470] [Q&A] [answered]
This is (I hope) an overly simple question, but I'm adding this to a more complex setup that already has go routines to keep things running. What would be the correct way to do that here? I'd want something that is linked to a context (that will be shared with other routines), e.g.

```go
go wait.UntilWithContext(ctx, startRiver, 0)

<-ctx.Done()
# clean up / close stuff
````

So my question would be - what should be in `StartRiver`? Do you have an example? Sorry if it's a very simple question - just looking at this library today (and very excited to use it)!

**Comment by vsoch** [comment]
Oh hmm looking at the bottom of https://github.com/riverqueue/river/blob/bf61772d4693ed498f66a7e1697819554bf3f0cb/client.go#L652. Do I just call Start and give it the context that I have (that I showed above)?

**Reply by brandur** [reply, marked as answer]
@vsoch Yep, all you need to do is to call `Client.Start` and the River client will run until one of the following conditions is met:

1. The given context is cancelled.
2. `Client.Stop` is called.

Generally (2) is preferred because by using `Stop` you give the client a chance to gracefully try to finish up with any jobs it's currently working. Cancelling the context will cause a "hard stop" that'll throw out existing work.

In your code above, you probably want to have a stopping signal that tries to stop gracefully before going full scorched earth and cancelling everything. e.g.

``` go
stopping := make(chan struct{})

// do start up and run

<-stopping
riverClient.Stop(ctx)
```

**Reply by vsoch** [reply]
Thank you! I’m not sure why but I didn’t get notified about this discussion reply - will need to look at my notifications (or subscribe when it is converted next time).
